File: shoope/746-min-cost-climbing-stairs/746.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

from typing import List


class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        dp: List[int] = [0 for i in range(len(cost) + 1)]
        for i in range(2, len(cost) + 1):
            dp[i] = min(dp[i - 2] + cost[i - 2], dp[i - 1] + cost[i - 1])

        return dp[-1]


if __name__ == '__main__':
    sol = Solution()
    print(sol.minCostClimbingStairs([10, 15, 20]))
    print(sol.minCostClimbingStairs([1, 100, 1, 1, 1, 100, 1, 1, 100, 1]))

File: shoope/746-min-cost-climbing-stairs/746.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        int size = cost.size();
        vector<int> dp = vector<int>(size+1, 0);
        for (int i = 2; i <= size; ++i) {
            dp[i] = min(dp[i-1] + cost[i-1], dp[i-2] + cost[i-2]);
        }
        return dp[size];
    }
};

int main(int argc, char** args) {
    Solution sol = Solution();
    vector<int> cost1 = {10, 15, 20};
    cout << sol.minCostClimbingStairs(cost1) << endl;
    
    vector<int> cost2 = {1, 100, 1, 1, 1, 100, 1, 1, 100, 1};
    cout << sol.minCostClimbingStairs(cost2) << endl;

    return 0;
}